tests/Unit/DataStructures/Test_ModalVector.cpp

```diff
@@ -0,0 +1,99 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include <array>
+#include <tuple>
+
+#include "DataStructures/ModalVector.hpp"    // IWYU pragma: keep
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/DereferenceWrapper.hpp"  // IWYU pragma: keep
+#include "Utilities/Functional.hpp"
+#include "Utilities/StdHelpers.hpp"  // IWYU pragma: keep
+#include "Utilities/TypeTraits.hpp"
+#include "tests/Unit/DataStructures/VectorImplTestHelper.hpp"
+
+// IWYU pragma: no_include "DataStructues/DenseMatrix.hpp"
+// IWYU pragma: no_include "tests/Utilities/MakeWithRandomValues.hpp"
+
+class DataVector;
+
+SPECTRE_TEST_CASE("Unit.DataStructures.ModalVector", "[DataStructures][Unit]") {
+  SECTION("test construct and assign") {
+    TestHelpers::VectorImpl::vector_test_construct_and_assign<ModalVector,
+                                                              double>();
+  }
+  SECTION("test serialize and deserialize") {
+    TestHelpers::VectorImpl::vector_test_serialize<ModalVector, double>();
+  }
+  TestHelpers::VectorImpl::vector_test_ref<ModalVector, double>();
+  TestHelpers::VectorImpl::vector_test_math_after_move<ModalVector, double>();
+}
+
+// [[OutputRegex, Must copy into same size]]
+[[noreturn]] SPECTRE_TEST_CASE("Unit.DataStructures.ModalVector.RefDiffSize",
+                               "[DataStructures][Unit]") {
+  ASSERTION_TEST();
+#ifdef SPECTRE_DEBUG
+  TestHelpers::VectorImpl::vector_ref_test_size_error<ModalVector>();
+  ERROR("Failed to trigger ASSERT in an assertion test");
+#endif
+}
+
+// [[OutputRegex, Must copy into same size]]
+[[noreturn]] SPECTRE_TEST_CASE(
```

Comment:
I think there's a third error test that does it for expressions of different size.